src/platform/Darwin/DnssdImpl.cpp

```diff
@@ -188,12 +218,88 @@ CHIP_ERROR Register(void * context, DnssdPublishCallback callback, uint32_t inte
     sdCtx = chip::Platform::New<RegisterContext>(type, callback, context);
     VerifyOrReturnError(nullptr != sdCtx, CHIP_ERROR_NO_MEMORY);
 
-    DNSServiceRef sdRef;
-    auto err = DNSServiceRegister(&sdRef, kRegisterFlags, interfaceId, name, type, kLocalDot, nullptr, ntohs(port), record.size(),
-                                  record.data(), OnRegister, sdCtx);
-    VerifyOrReturnError(kDNSServiceErr_NoError == err, sdCtx->Finalize(err));
+    // Treat empty hostname as null.
+    if (hostname != nullptr && strcmp(hostname, "") == 0)
+    {
+        hostname = nullptr;
+    }
+    std::string hostnameStr;
+    if (hostname != nullptr)
+    {
+        hostnameStr = std::string(hostname) + '.' + kLocalDot;
+        hostname    = hostnameStr.c_str();
+    }
 
-    return MdnsContexts::GetInstance().Add(sdCtx, sdRef);
+    DNSServiceRef sdRef = nullptr;
+    if (hostname)
+    {
+        auto err = DNSServiceCreateConnection(&sdRef);
+        if (sdRef != nullptr && !err)
+        {
+            // Get all the interface addresses
+            struct ifaddrs * ifaddr;
+            auto ret = getifaddrs(&ifaddr);
+            if (ret)
+            {
+                ChipLogError(Discovery, "Getting interface addresses failed %d. Proceed registering service with default hostname",
+                             ret);
+                freeifaddrs(ifaddr);
+                return RegisterService(sdRef, kRegisterFlags, interfaceId, name, type, kLocalDot, nullptr, port, record.size(),
+                                       record.data(), OnRegister, sdCtx);
+            }
+
+            // Call DNSServiceRegisterRecord for each interface so we can register the address for each interface
```

Comment:
Do we want to register the "hostname" for all interfaces if the interface has been specified using the command line flag ? So for cases where the interface is not `kDNSServiceInterfaceIndexAny`

Reply:
yes that makes sense. Can you please point me to the command line usage to see what parameters can be passed in?

Reply:
done